apps/heft/src/utilities/fileSystem/TypeScriptCachedFileSystem.ts

```diff
@@ -0,0 +1,235 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import * as fs from 'fs';
+import * as nodeJsPath from 'path';
+import {
+  Encoding,
+  Text,
+  IFileSystemWriteFileOptions,
+  IFileSystemReadFileOptions,
+  IFileSystemCopyFileOptions,
+  IFileSystemDeleteFileOptions,
+  IFileSystemCreateLinkOptions,
+  FileSystem,
+  FileSystemStats
+} from '@rushstack/node-core-library';
+
+export interface IReadFolderFilesAndDirectoriesResult {
+  files: string[];
+  directories: string[];
+}
+
+export interface ICreateHardLinkExtendedOptions extends IFileSystemCreateLinkOptions {
+  preserveExisting: boolean;
+}
+
+interface ICacheEntry<TEntry> {
+  entry?: TEntry;
```

Comment:
Would `TEntry | undefined` be better here? It's not really "optional", it's possibly undefined.